### ISSUE TITLE
Count hanging (not single cell)

### ISSUE THREAD

**Issue by bdoughty** [CGATOxford/UMI-tools#241]
I have been happily using UMItools for single cell experiments for a while, but am now trying to analyze bulk data without cell barcodes. I have followed the procedure outlined in the quick start guide, but am running into trouble when counting. I have about 25.5M reads (of which ~85% map), and when I run umi_tools count on it, I get the following output:

# pid: 8256, system: Linux 2.6.32-696.20.1.el6.x86_64 #1 SMP Fri Jan 12 15:07:59 EST 2018 x86_64
# cell_tag                                : None
# cell_tag_delim                          : None
# cell_tag_split                          : -
# chrom                                   : None
# compresslevel                           : 6
# gene_tag                                : None
# gene_transcript_map                     : None
# get_umi_method                          : read_id
# ignore_umi                              : False
# in_sam                                  : False
# log2stderr                              : False
# loglevel                                : 1
# mapping_quality                         : 0
# method                                  : directional
# paired                                  : False
# per_cell                                : False
# per_contig                              : True
# per_gene                                : True
# random_seed                             : None
# short_help                              : None
# skip_regex                              : ^[__|Unassigned]
# stderr                                  : <_io.TextIOWrapper name='<stderr>' mode='w' encoding='ANSI_X3.4-1968'>
# stdin                                   : <_io.TextIOWrapper name='normbulk_S19_R2_001.extracted.sorted.bam' mode='r' encoding='ANSI_X3.4-1968'>
# stdlog                                  : <_io.TextIOWrapper name='normbulk_S19.count.log' mode='a' encoding='ANSI_X3.4-1968'>
# stdout                                  : <_io.TextIOWrapper name='normbulk_S19_counts.tsv' mode='w' encoding='ANSI_X3.4-1968'>
# subset                                  : None
# threshold                               : 1
# timeit_file                             : None
# timeit_header                           : None
# timeit_name                             : all
# umi_sep                                 : _
# umi_tag                                 : RX
# umi_tag_delim                           : None
# umi_tag_split                           : None
# wide_format_cell_counts                 : False
2018-03-08 12:17:31,105 INFO Parsed 1000000 input reads
2018-03-08 12:17:31,106 INFO Parsed 2000000 input reads
2018-03-08 12:32:47,338 INFO Parsed 3000000 input reads
2018-03-08 12:32:47,339 INFO Parsed 4000000 input reads
...
2018-03-08 12:32:47,341 INFO Parsed 16000000 input reads
2018-03-08 12:32:47,341 INFO Parsed 17000000 input reads

Basically, it quickly processes 2M reads, then waits a while, then processes 15M more, then hangs indefinitely, until the job is killed by our scheduler. The same thing happens when I just try to dedup instead of counting. Do you have any idea what the problem could be? Thanks.

**Comment by TomSmithCGAT** [comment]
Hi @bdoughty,

My guess is that first ~2M reads can be easily processed and then it reaches a position with a lot of unique UMIs which takes 15 minutes to process. This happens again around 15M reads where there are so many reads that the process of building the network and resolving it requires too much memory and the job gets killed. 

You say you're analysing bulk RNA-Seq data. In this case, you should use `umi_tools dedup` rather than `umi_tools count`. Assuming you have use a standard bulk RNA-Seq library preparation method where fragmentation occurs prior to amplification, the process you want to perform is `umi_tool extract` -> aligner -> `umi_tools dedup` -> counter, where the aligner is up to you and can be either to the genome or transcriptome, and the counter is either e.g `featureCounts` if you have aligned to the genome, or a simple script to count reads per contig if you have aligned to the transcriptome. From your use of the `--per-contig` option above, I assume you have aligned to the transcriptome?

`umi_tools count` is hardcoded to use the `--per-gene` options which means all reads per gene are treated as possible duplicates, irrespective of their position on the gene. This behaviour is designed to deal with library prep. methods where reads are expected to originate from the 3' end of the read and fragmentation occurs post amplification, thus the position of the read in the gene cannot be used to determine duplicate reads. If applied to bulk RNA-seq, this will massively inflate the number of unique UMIs which have to be resolved per position (gene) leading to the outcome above.

All the above is based on assumptions I've made from what you've said so please correct me if I've missunderstood at all.

**Comment by TomSmithCGAT** [comment]
@IanSudbery - I'm going to add this to the next version issues as a reminder to update the documentation to make it clearer than `umi_tools count` is only applicable for scRNA-Seq and how to count reads/gene in bulk RNA-Seq.

**Comment by bdoughty** [comment]
Hi @TomSmithCGAT,

Thanks for the prompt answer. Just to clarify a few things about our experiment in response to your questions:
1) We're not doing a standard RNAseq library prep scheme with pA capture, amplification, and fragmentation, but instead amplifying selectively (<100 unique sequences) from the transcriptome and sequencing. These sequences should always come from the same position on their respective transcripts, and we add a UMI (and optionally a cell barcode, if doing single cell) prior to amplification.
2) The workflow for single cell was umi_tools extract -> align (to transcriptome that I built that has only the relevant portions of the subset of genes we care about) -> umi_tools count -> downstream analysis on the cells by genes table. I was hoping to do something similar for the bulk data, and didn't realize that umi_tools count was only for single cell (but I can write a simple script to take the output of dedup).
3) However, when I ran umi_tools dedup on the same bam file, basically the same thing happened. I ran dedup with no special flags, and it quickly took care of the first 2M, then waited 20 minutes for the next 15M, then hung for 12 hours before finishing the remaining 4.5M reads. I have pasted the stats below, and it's notable is that we went from 25M reads to just 45K after deduping. This number is lower than what we get if we do a naive dedup (deduping based on exact identity and then throwing out low abundance reads gives around 180K, which I guess might not be that off come to think of it, but it's still worrisome that it took 12hr).

2018-03-08 21:15:45,945 INFO Parsed 1000000 input reads
2018-03-08 21:15:45,946 INFO Parsed 2000000 input reads
2018-03-08 21:35:41,629 INFO Parsed 3000000 input reads
2018-03-08 21:35:41,631 INFO Parsed 4000000 input reads
...
2018-03-08 21:35:41,633 INFO Parsed 16000000 input reads
2018-03-08 21:35:41,633 INFO Parsed 17000000 input reads
2018-03-09 09:03:04,782 INFO Parsed 18000000 input reads
2018-03-09 09:03:04,784 INFO Parsed 19000000 input reads
2018-03-09 09:03:04,784 INFO Parsed 20000000 input reads
2018-03-09 09:03:04,784 INFO Parsed 21000000 input reads
2018-03-09 09:27:07,092 INFO Reads: Input Reads: 21666507
2018-03-09 09:27:07,096 INFO Number of reads out: 46163
2018-03-09 09:27:07,096 INFO Total number of positions deduplicated: 42
2018-03-09 09:27:07,096 INFO Mean number of unique UMIs per position: 15569.62
2018-03-09 09:27:07,097 INFO Max. number of unique UMIs per position: 467348

(Also, what is a "unique UMI" in this context? I'm not sure I understand what it means for the mean number of unique UMIs per position to be 15.5K.)

4) Now, I understand why treating all reads to a specific gene as potential duplicates would get expensive for bulk RNA-seq, and I'm wondering if the problem is that all of our sequences start and end at the same position (i.e. for a specific gene, all the reads will be at the same coordinates, since we don't do any fragmentation). If the problem is that potentially everything mapping to one gene can be a PCR duplicate, is there any solution to how long this takes?
5) Just FYI, we are using 16bp UMIs, and I didn't change the Hamming Distance threshold.

I guess my main question is, since our library prep scheme is different, is this always going to happen when analyzing bulk data?

**Comment by bdoughty** [comment]
Follow-up question:
Where do the stats at the bottom of a dedup/group log come from? I am a little confused because I got the following results from deduping a small test set of reads (1M before mapping):

2018-03-26 13:24:31,173 INFO total_umis 719208
2018-03-26 13:24:31,174 INFO #umis 262157
2018-03-26 14:25:25,061 INFO Written out 100000 reads
2018-03-26 14:54:54,638 INFO Reads: Input Reads: 719208
2018-03-26 14:54:54,639 INFO Number of reads out: 158738
2018-03-26 14:54:54,639 INFO Total number of positions deduplicated: 113
2018-03-26 14:54:54,639 INFO Mean number of unique UMIs per position: 2320.18
2018-03-26 14:54:54,639 INFO Max. number of unique UMIs per position: 26950

How are these numbers generated? How are there ~1.5 more UMIs than output reads, and why is the product of the number of positions and the mean unique UMIs/position not quite equal to the number of unique umis? Also, is there anything to be done about our experimental design where there are only a few coordinates where our reads should map to, or will using UMItools just take a long time? Are some of the other clustering options better than directional, in this case?

Thanks very much!

**Comment by TomSmithCGAT** [comment]
I'm going to close this issue due to low activity. Please re-open if the issue has not been resolved
